### PR TITLE
[WIP] remote: use explicit tree parameter for cache operations

### PR DESCRIPTION
**Pull request by pmrowla**
* [ ] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [ ] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here. If the CLI API is changed, I have updated [tab completion scripts](https://github.com/iterative/dvc/tree/master/scripts/completion).

* [ ] ❌ I will check DeepSource, CodeClimate, and other sanity checks below. (We consider them recommendatory and don't expect everything to be addressed. Please fix things that actually improve code or fix bugs.)

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

* `remote.save()` now takes an explicit `tree` parameter
    * When `tree` is `remote.tree`, save is done via `move()` + `link()`
    * For all other trees, save is done by copying object from `tree` into `remote.tree`

### CODE REVIEW

File: dvc/remote/base.py [hunk 1]
```diff
@@ -397,51 +400,44 @@ def supported(cls, config):
...
-    def _calculate_checksums(self, file_infos):
+    def _calculate_checksums(self, file_infos, tree):
         file_infos = list(file_infos)
         with Tqdm(
             total=len(file_infos),
             unit="md5",
             desc="Computing file/dir hashes (only done once)",
         ) as pbar:
-            worker = pbar.wrap_fn(self.get_file_checksum)
+            worker = pbar.wrap_fn(tree.get_file_checksum)
```
Comment (efiop) [inline dvc/remote/base.py:410]: So we still don't have a cache/remote separation?

Comment (pmrowla) [reply dvc/remote/base.py:410]: @efiop Not yet. I'm still thinking about the best way to go about that particular separation.

The "remote" functions like `push`/`pull`/`status` (which are currently in `LocalRemote`) are really functions for syncing the contents of two caches. So I've been wondering if they should just be abstracted into functions for syncing from "this/self" cache to/from any other cache, and then moved into `BaseRemote`?

And then the remote classes could still be renamed to `Cache` or `RemoteCache` (or something else along those lines), but there wouldn't be any further separation into new classes (other than the tree methods).

Comment (pmrowla) [reply dvc/remote/base.py:410]: I know that we would still have the external output/dependency use case to consider, but it seems like in practice, that is still just a situation where we would have two "s3 caches"

* one which is used as the typical "dvc remote" for syncing local cache <-> s3
* and the second which is used as the s3 external out cache

**Comment by pmrowla** [comment]
closed in favor of #4019

**Pull request closed without merging**